File: 2962. Count Subarrays Where Max Element Appears at Least K Times/countSubarrays.py
from typing import List

class Solution:
  def countSubarrays(self, nums: List[int], k: int) -> int:
    max_n, max_cnt = max(nums), 0
    left, res = 0, 0
    
    for right in range(len(nums)):
      if nums[right] == max_n:
        max_cnt += 1
        
      while max_cnt > k or (left <= right and max_cnt == k and nums[left] != max_n):
        if nums[left] == max_n:
          max_cnt -= 1
        left += 1
        
      if max_cnt == k:
        res += left + 1
        
    return res

nums = [1, 3, 2, 3, 3]
k = 2
solution = Solution()
print('countSubarrays: ', solution.countSubarrays(nums, k))
          

File: 2962. Count Subarrays Where Max Element Appears at Least K Times/countSubarrays.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const countSubarrays = (nums, k) => {
  let maxN = Math.max(...nums);
  let maxCnt = 0;

  let left = 0;
  let res = 0;

  for (let right = 0; right < nums.length; right++) {
    if (nums[right] === maxN) {
      maxCnt++;
    }

    while (maxCnt > k || (left <= right && maxCnt === k && nums[left] !== maxN)) {
      if (nums[left] === maxN) {
        maxCnt--;
      }
      left++;
    }

    if (maxCnt === k) {
      res += left + 1
    }
  }

  return res;
};

const nums = [1, 3, 2, 3, 3];
const k = 2;
console.log('countSubarrays: ', countSubarrays(nums, k));
